### PR TITLE
mrs: Make thread safe and related cleanups

### PR DESCRIPTION
**Pull request by bsdjhb**
- mrs: Add utrace(2) events
- mrs: Use %#p instead of mrs_printcap
- mrs: Various style fixes, no functional change
- mrs: Apply recent realloc fixes to mrs_rallocx
- mrs: Make thread safe

### CODE REVIEW

**Comment by bsdjhb** [comment]
This fixes my `git grep` crashes as well as most of the crashes I was seeing with the GUI.  I can now boot and login to wayland plasma just fine and start a konsole without any crash.

Here are the traces I got via the new utrace events showing that mrs was doing a double-free to the backend allocator for the `git grep` crash:

```
> kdump -H -t u | grep 0x45e00000
 21593 101243 git      USER  0x45e00000 = malloc(50708)
 21593 101243 git      USER  0x45e00000 = mrs_malloc(50708)
 21593 101243 git      USER  mrs_free(0x45e00000)
 21593 101243 git      USER  quarantine_insert(0x45e00000, 57344)
 21593 101242 git      USER  free(0x45e00000)
 21593 101075 git      USER  free(0x45e00000)
```

Note separate threads both invoking the backend allocator's free for this address.  Other entries in the ktrace showed both threads also calling the cheri_revoke system calls interleaved with each other.

**Comment by jrtc27** [comment]
check_and_perform_flush isn't under a lock, so isn't there a risk we do a second revocation sweep right after the first one's done if we race there? I suspect you want check_and_perform_flush to do a lock-free check whether it should revoke and then check again under the lock.

File: lib/libc/stdlib/malloc/mrs/mrs.c [hunk 1]
```diff
@@ -280,30 +300,48 @@ static struct mrs_quarantine offload_quarantine; /* quarantine for the offload t
...
-	pthread_mutex_t name; \
-	char name ## _buf[256] __attribute__((aligned(16))); \
-	void *name ## _storage() { \
-		return name ## _buf; \
+#define create_lock(name)						\
+	pthread_mutex_t name;						\
+	char name ## _buf[256] __attribute__((aligned(16)));		\
+									\
+	void *								\
+	name ## _storage()						\
```
Comment (brooksdavis) [inline lib/libc/stdlib/malloc/mrs/mrs.c:336]:
```suggestion
	name ## _storage(void)						\
```
I think newer compilers will whine about this otherwise?

Comment (bsdjhb) [reply lib/libc/stdlib/malloc/mrs/mrs.c:336]: Oof, yes.  I fixed several of those in other places but missed this one.

Comment (bsdjhb) [reply lib/libc/stdlib/malloc/mrs/mrs.c]: Oh, it's actually terrible.  The function takes two size_t arguments and fixing this actually breaks the build.  Have to add the actual args.

**Comment by bsdjhb** [comment]
Ok, this version should be less terrible in the face of concurrency, and in a ktrace of `git grep` I can see other threads able to malloc() and free() concurrently with a thread flushing the quarantine.  I've used the second strategy of taking a snapshot of the application quarantine under the lock and resetting it to empty.  Then dropping the lock and doing the revoke on the local copy.  This does mean you can overshoot your quarantine limit, in particular in the offload case where they have to synchronize with the offload thread to wait it to drain, so in the worst case you could have one full quarantine per thread all queued up while waiting for the offload thread to drain.  I kind of think this means the offload thread should just go away most likely.  Note that this approach is also quite close to just having a per-thread application quarantine and at this point the code changes to support that would be relatively modest if we wanted to research that idea.  A naive per-thread application quarantine would suffer from the same quarantine limit overshoot problem FWIW (or rather, the limit is really a per-thread limit at that point rather than a global limit).

File: lib/libc/stdlib/malloc/mrs/mrs.c [hunk 2]
```diff
@@ -63,36 +63,47 @@
...
- * DEBUG: print debug statements
- * PRINT_STATS: print statistics on exit
- * PRINT_CAPREVOKE: print stats for each CHERI revocation
- * PRINT_CAPREVOKE_MRS: print details of MRS operation around revocations
- * CLEAR_ON_ALLOC: zero allocated regions as they are allocated (for non-calloc allocation functions)
- * CLEAR_ON_RETURN: zero allocated regions as they come out of quarantine
- * CLEAR_ON_FREE: zero allocated regions as they are given to us
- * REVOKE_ON_FREE: perform revocation on free rather than during allocation routines
+ *   If !OFFLOAD_QUARANTINE, this still prevents the application from
+ *   using CPU 2.
```
Comment (nwf) [inline lib/libc/stdlib/malloc/mrs/mrs.c:76]: That probably should be "CPU 3", while you're here.

File: lib/libc/stdlib/malloc/mrs/mrs.c [hunk 3]
```diff
@@ -1620,28 +1620,23 @@ mrs_rallocx(void *ptr, size_t size, int flags)
...
-		return (mrs_calloc(1, size));
-	}
-
 	void *new_alloc = mrs_calloc(1, size);
-	/* old object is not deallocated according to the spec */
-	if (new_alloc == NULL) {
-		return (NULL);
+
+	/*
+	 * per the C standard, copy and free IFF the old pointer is valid
```
Comment (brooksdavis) [inline lib/libc/stdlib/malloc/mrs/mrs.c:1650]: Capitalize "Per" to match other comment now that it's updated.

**Review by brooksdavis — APPROVED** [review]

**Pull request merged**